# CM4125 Week 3: Data Manipulation

In [ ]:
# This cell is used to change parameter of the rise slideshow, 
# such as the window width/height and enabling a scroll bar
from notebook.services.config import ConfigManager
cm = ConfigManager()
cm.update('livereveal', {
              'width': 1700,
              'height': 800,
              'scroll': True,
})
# This code lets you colour the axis lines white
from matplotlib import style
style.use('dark_background')
# This code allows you to show images within the notebook
%matplotlib inline

## Coursework Clarification!
Today @ 4:00 pm, I will open a Zoom room

## Lecture objectives

1) Load a set of datasets containing different characteristics

2) Understand the best methods to manipulate it prior to analysis

3) Apply more elaborated pre-processing techniques to clean data

## Loading data (revised)

If you recall, last week we had to download and upload csv files

This is obviously not very practical!

The `Pandas` module allows us to load data from csv files that are stored online in places such as:

    Github
    Dropbox
    AWS
    etc

In [1]:
# Importing a dataset that I have in dropbox
import pandas as pd
# Copy dropbox link, change dl=0 for raw=1
url = 'https://www.dropbox.com/s/ju1iemc8k1wp0p5/time.csv?raw=1'
time = pd.read_csv(url)
time 

Year               Honor                Name        Country  Birth Year  \
0   1927     Man of the Year   Charles Lindbergh  United States      1902.0   
1   1928     Man of the Year     Walter Chrysler  United States      1875.0   
2   1929     Man of the Year       Owen D. Young  United States      1874.0   
3   1930     Man of the Year      Mahatma Gandhi          India      1869.0   
4   1931     Man of the Year        Pierre Laval         France      1883.0   
..   ...                 ...                 ...            ...         ...   
86  2012  Person of the Year        Barack Obama  United States      1961.0   
87  2013  Person of the Year        Pope Francis   Vatican City      1936.0   
88  2014  Person of the Year  The Ebola Fighters            NaN         NaN   
89  2015  Person of the Year       Angela Merkel        Germany      1954.0   
90  2016  Person of the Year        Donald Trump  United States      1946.0   

    Death Year                                              Title    Category  \
0       1974.0                                  US Air Mail Pilot         NaN   
1       1940.0                                Founder of Chrysler   Economics   
2       1962.0  Member of the German Reparations International...   Diplomacy   
3       1948.0                                                NaN  Revolution   
4       1945.0                           Prime Minister of France    Politics   
..         ...                                                ...         ...   
86         NaN                     President of the United States    Politics   
87         NaN                  Pope of the Roman Catholic Church    Religion   
88         NaN                                                NaN     Science   
89         NaN                              Chancellor of Germany    Politics   
90         NaN                     President of the United States    Politics   

                                              Context  
0                     First Solo Transatlantic Flight  
1                               Chrysler/Dodge Merger  
2                                          Young Plan  
3                                          Salt March  
4                                                 NaN  
..                                                ...  
86                              Presidential Election  
87                                     Papal Conclave  
88                                     Ebola Epidemic  
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...  
90                              Presidential Election  

[91 rows x 9 columns]

This dataset contains the list of people named Man/Person of the Year by Times magazine

## Data by Columns

Last week we added new columns to dataframes

    The dog/cat class at the end of the image repository
    The "watched" column at the end of the Netflix dataset

Today, we will learn other ways in which we can manipulate, add or relate data from columns

### Mapping Columns

Sometimes we want to convert a column of categorical data from strings (text) to integers (numbers)

We can define a `dictionary` which maps the strings (as keys) to integers (as values) using the `map` method

Another option is to define a `set` of the unique categories as follows.

In [2]:
categories = set(time['Category'])
categories

{'Diplomacy',
 'Economics',
 'Environment',
 'Media',
 'Philanthropy',
 'Politics',
 'Religion',
 'Revolution',
 'Science',
 'Space',
 'Technology',
 'War',
 nan}

Then, we can then assign unique numbers to this `set` as a `dictionary` (defined in Python using the `{}`)

In [3]:
mapping = {category : index for index, category in enumerate(categories)}
mapping

{nan: 0,
 'Environment': 1,
 'Philanthropy': 2,
 'Politics': 3,
 'Revolution': 4,
 'Religion': 5,
 'Technology': 6,
 'Economics': 7,
 'War': 8,
 'Media': 9,
 'Diplomacy': 10,
 'Space': 11,
 'Science': 12}

We now have a `map` which allows us to convert from a category string to an `int`. For example:

In [4]:
mapping['Diplomacy']

10

In [5]:
mapping['Media']

9

We can use the `map` method on the chosen column to do the conversion

The `map` method takes our `dictionary` as an argument and returns a new column with the mapping applied

In [6]:
time['Category'].map(mapping)

0      0
1      7
2     10
3      4
4      3
      ..
86     3
87     5
88    12
89     3
90     3
Name: Category, Length: 91, dtype: int64

If we wanted to overwrite the `Category` column with our new mapped values, we could do:

    time['Category'] = time['Category'].map(mapping)

However, it may be best to keep both. So we can instead assign it a different column name

In [7]:
time['Category Ordinal'] = time['Category'].map(mapping)
time

Year               Honor                Name        Country  Birth Year  \
0   1927     Man of the Year   Charles Lindbergh  United States      1902.0   
1   1928     Man of the Year     Walter Chrysler  United States      1875.0   
2   1929     Man of the Year       Owen D. Young  United States      1874.0   
3   1930     Man of the Year      Mahatma Gandhi          India      1869.0   
4   1931     Man of the Year        Pierre Laval         France      1883.0   
..   ...                 ...                 ...            ...         ...   
86  2012  Person of the Year        Barack Obama  United States      1961.0   
87  2013  Person of the Year        Pope Francis   Vatican City      1936.0   
88  2014  Person of the Year  The Ebola Fighters            NaN         NaN   
89  2015  Person of the Year       Angela Merkel        Germany      1954.0   
90  2016  Person of the Year        Donald Trump  United States      1946.0   

    Death Year                                              Title    Category  \
0       1974.0                                  US Air Mail Pilot         NaN   
1       1940.0                                Founder of Chrysler   Economics   
2       1962.0  Member of the German Reparations International...   Diplomacy   
3       1948.0                                                NaN  Revolution   
4       1945.0                           Prime Minister of France    Politics   
..         ...                                                ...         ...   
86         NaN                     President of the United States    Politics   
87         NaN                  Pope of the Roman Catholic Church    Religion   
88         NaN                                                NaN     Science   
89         NaN                              Chancellor of Germany    Politics   
90         NaN                     President of the United States    Politics   

                                              Context  Category Ordinal  
0                     First Solo Transatlantic Flight                 0  
1                               Chrysler/Dodge Merger                 7  
2                                          Young Plan                10  
3                                          Salt March                 4  
4                                                 NaN                 3  
..                                                ...               ...  
86                              Presidential Election                 3  
87                                     Papal Conclave                 5  
88                                     Ebola Epidemic                12  
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...                 3  
90                              Presidential Election                 3  

[91 rows x 10 columns]

### Manipulating Numerical Columns

We can perform arithmetic between columns as if they were simply numbers

We will take the column `Death Year`, subtract the column `Birth Year`, and storing the result in a (new) column called `Lifespan`

In [8]:
time['Lifespan'] = time['Death Year'] - time['Birth Year']
time

Year               Honor                Name        Country  Birth Year  \
0   1927     Man of the Year   Charles Lindbergh  United States      1902.0   
1   1928     Man of the Year     Walter Chrysler  United States      1875.0   
2   1929     Man of the Year       Owen D. Young  United States      1874.0   
3   1930     Man of the Year      Mahatma Gandhi          India      1869.0   
4   1931     Man of the Year        Pierre Laval         France      1883.0   
..   ...                 ...                 ...            ...         ...   
86  2012  Person of the Year        Barack Obama  United States      1961.0   
87  2013  Person of the Year        Pope Francis   Vatican City      1936.0   
88  2014  Person of the Year  The Ebola Fighters            NaN         NaN   
89  2015  Person of the Year       Angela Merkel        Germany      1954.0   
90  2016  Person of the Year        Donald Trump  United States      1946.0   

    Death Year                                              Title    Category  \
0       1974.0                                  US Air Mail Pilot         NaN   
1       1940.0                                Founder of Chrysler   Economics   
2       1962.0  Member of the German Reparations International...   Diplomacy   
3       1948.0                                                NaN  Revolution   
4       1945.0                           Prime Minister of France    Politics   
..         ...                                                ...         ...   
86         NaN                     President of the United States    Politics   
87         NaN                  Pope of the Roman Catholic Church    Religion   
88         NaN                                                NaN     Science   
89         NaN                              Chancellor of Germany    Politics   
90         NaN                     President of the United States    Politics   

                                              Context  Category Ordinal  \
0                     First Solo Transatlantic Flight                 0   
1                               Chrysler/Dodge Merger                 7   
2                                          Young Plan                10   
3                                          Salt March                 4   
4                                                 NaN                 3   
..                                                ...               ...   
86                              Presidential Election                 3   
87                                     Papal Conclave                 5   
88                                     Ebola Epidemic                12   
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...                 3   
90                              Presidential Election                 3   

    Lifespan  
0       72.0  
1       65.0  
2       88.0  
3       79.0  
4       62.0  
..       ...  
86       NaN  
87       NaN  
88       NaN  
89       NaN  
90       NaN  

[91 rows x 11 columns]

Notice that the method acknowledged that some columns were `NaN` (since the person is still alive) and thus put a `NaN` as a result in the lifespan column.

Sometimes you want to do "more difficult" numerical conversions

In those cases it is more convenient to define a **function** and then use the `apply` method

In this example, we will define a function to convert years to decades to be applied in the different columns that contain years

In [9]:
def to_decade(value):
    return 10 * (value // 10)

First we should test our function on some examples to ensure the function is correct.

In [11]:
to_decade(1971)

1970

Now we can use `apply` to transform the entire column `Birth Year` into the corresponding decade

In [12]:
time['Birth Decade'] = time['Birth Year'].apply(to_decade)
time

Year               Honor                Name        Country  Birth Year  \
0   1927     Man of the Year   Charles Lindbergh  United States      1902.0   
1   1928     Man of the Year     Walter Chrysler  United States      1875.0   
2   1929     Man of the Year       Owen D. Young  United States      1874.0   
3   1930     Man of the Year      Mahatma Gandhi          India      1869.0   
4   1931     Man of the Year        Pierre Laval         France      1883.0   
..   ...                 ...                 ...            ...         ...   
86  2012  Person of the Year        Barack Obama  United States      1961.0   
87  2013  Person of the Year        Pope Francis   Vatican City      1936.0   
88  2014  Person of the Year  The Ebola Fighters            NaN         NaN   
89  2015  Person of the Year       Angela Merkel        Germany      1954.0   
90  2016  Person of the Year        Donald Trump  United States      1946.0   

    Death Year                                              Title    Category  \
0       1974.0                                  US Air Mail Pilot         NaN   
1       1940.0                                Founder of Chrysler   Economics   
2       1962.0  Member of the German Reparations International...   Diplomacy   
3       1948.0                                                NaN  Revolution   
4       1945.0                           Prime Minister of France    Politics   
..         ...                                                ...         ...   
86         NaN                     President of the United States    Politics   
87         NaN                  Pope of the Roman Catholic Church    Religion   
88         NaN                                                NaN     Science   
89         NaN                              Chancellor of Germany    Politics   
90         NaN                     President of the United States    Politics   

                                              Context  Category Ordinal  \
0                     First Solo Transatlantic Flight                 0   
1                               Chrysler/Dodge Merger                 7   
2                                          Young Plan                10   
3                                          Salt March                 4   
4                                                 NaN                 3   
..                                                ...               ...   
86                              Presidential Election                 3   
87                                     Papal Conclave                 5   
88                                     Ebola Epidemic                12   
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...                 3   
90                              Presidential Election                 3   

    Lifespan  Birth Decade  
0       72.0        1900.0  
1       65.0        1870.0  
2       88.0        1870.0  
3       79.0        1860.0  
4       62.0        1880.0  
..       ...           ...  
86       NaN        1960.0  
87       NaN        1930.0  
88       NaN           NaN  
89       NaN        1950.0  
90       NaN        1940.0  

[91 rows x 12 columns]

### Manipulating String Columns

`apply` can also be used on strings

For example, we can apply the `len` function to `Name` to get the number of characters in each name.

In [13]:
time['Name Length'] = time['Name'].apply(len)
time

Year               Honor                Name        Country  Birth Year  \
0   1927     Man of the Year   Charles Lindbergh  United States      1902.0   
1   1928     Man of the Year     Walter Chrysler  United States      1875.0   
2   1929     Man of the Year       Owen D. Young  United States      1874.0   
3   1930     Man of the Year      Mahatma Gandhi          India      1869.0   
4   1931     Man of the Year        Pierre Laval         France      1883.0   
..   ...                 ...                 ...            ...         ...   
86  2012  Person of the Year        Barack Obama  United States      1961.0   
87  2013  Person of the Year        Pope Francis   Vatican City      1936.0   
88  2014  Person of the Year  The Ebola Fighters            NaN         NaN   
89  2015  Person of the Year       Angela Merkel        Germany      1954.0   
90  2016  Person of the Year        Donald Trump  United States      1946.0   

    Death Year                                              Title    Category  \
0       1974.0                                  US Air Mail Pilot         NaN   
1       1940.0                                Founder of Chrysler   Economics   
2       1962.0  Member of the German Reparations International...   Diplomacy   
3       1948.0                                                NaN  Revolution   
4       1945.0                           Prime Minister of France    Politics   
..         ...                                                ...         ...   
86         NaN                     President of the United States    Politics   
87         NaN                  Pope of the Roman Catholic Church    Religion   
88         NaN                                                NaN     Science   
89         NaN                              Chancellor of Germany    Politics   
90         NaN                     President of the United States    Politics   

                                              Context  Category Ordinal  \
0                     First Solo Transatlantic Flight                 0   
1                               Chrysler/Dodge Merger                 7   
2                                          Young Plan                10   
3                                          Salt March                 4   
4                                                 NaN                 3   
..                                                ...               ...   
86                              Presidential Election                 3   
87                                     Papal Conclave                 5   
88                                     Ebola Epidemic                12   
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...                 3   
90                              Presidential Election                 3   

    Lifespan  Birth Decade  Name Length  
0       72.0        1900.0           17  
1       65.0        1870.0           15  
2       88.0        1870.0           13  
3       79.0        1860.0           14  
4       62.0        1880.0           12  
..       ...           ...          ...  
86       NaN        1960.0           12  
87       NaN        1930.0           12  
88       NaN           NaN           18  
89       NaN        1950.0           13  
90       NaN        1940.0           12  

[91 rows x 13 columns]

Another example, we can create a function to get the initials from a name.

In [14]:
# First we define the function
def to_initials(name):
    import numpy as np # we need this for Python to know what NaN is!
    if name == np.NaN:
        return np.NaN
    else:
        initials = ""
        for word in name.split(' '):
            first_letter = word[0]
            initials += first_letter
        return initials
    # Then, we try it
to_initials("Barack Obama")

'BO'

Applying this gives us a name column with the person's initials.

In [15]:
time['Initials'] = time['Name'].apply(to_initials)
time

Year               Honor                Name        Country  Birth Year  \
0   1927     Man of the Year   Charles Lindbergh  United States      1902.0   
1   1928     Man of the Year     Walter Chrysler  United States      1875.0   
2   1929     Man of the Year       Owen D. Young  United States      1874.0   
3   1930     Man of the Year      Mahatma Gandhi          India      1869.0   
4   1931     Man of the Year        Pierre Laval         France      1883.0   
..   ...                 ...                 ...            ...         ...   
86  2012  Person of the Year        Barack Obama  United States      1961.0   
87  2013  Person of the Year        Pope Francis   Vatican City      1936.0   
88  2014  Person of the Year  The Ebola Fighters            NaN         NaN   
89  2015  Person of the Year       Angela Merkel        Germany      1954.0   
90  2016  Person of the Year        Donald Trump  United States      1946.0   

    Death Year                                              Title    Category  \
0       1974.0                                  US Air Mail Pilot         NaN   
1       1940.0                                Founder of Chrysler   Economics   
2       1962.0  Member of the German Reparations International...   Diplomacy   
3       1948.0                                                NaN  Revolution   
4       1945.0                           Prime Minister of France    Politics   
..         ...                                                ...         ...   
86         NaN                     President of the United States    Politics   
87         NaN                  Pope of the Roman Catholic Church    Religion   
88         NaN                                                NaN     Science   
89         NaN                              Chancellor of Germany    Politics   
90         NaN                     President of the United States    Politics   

                                              Context  Category Ordinal  \
0                     First Solo Transatlantic Flight                 0   
1                               Chrysler/Dodge Merger                 7   
2                                          Young Plan                10   
3                                          Salt March                 4   
4                                                 NaN                 3   
..                                                ...               ...   
86                              Presidential Election                 3   
87                                     Papal Conclave                 5   
88                                     Ebola Epidemic                12   
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...                 3   
90                              Presidential Election                 3   

    Lifespan  Birth Decade  Name Length Initials  
0       72.0        1900.0           17       CL  
1       65.0        1870.0           15       WC  
2       88.0        1870.0           13      ODY  
3       79.0        1860.0           14       MG  
4       62.0        1880.0           12       PL  
..       ...           ...          ...      ...  
86       NaN        1960.0           12       BO  
87       NaN        1930.0           12       PF  
88       NaN           NaN           18       TE  
89       NaN        1950.0           13       AM  
90       NaN        1940.0           12       DT  

[91 rows x 14 columns]

### Removing Columns (Unwanted Variables)

There are two main ways to remove them, either we specify which columns we want to **keep** or we specify which we want to **remove**

Here is the version where we specify, by name, which columns to **keep**. You saw this in an earlier lecture as selecting columns.

In [16]:
time[['Year', 'Honor', 'Name']]

Year               Honor                Name
0   1927     Man of the Year   Charles Lindbergh
1   1928     Man of the Year     Walter Chrysler
2   1929     Man of the Year       Owen D. Young
3   1930     Man of the Year      Mahatma Gandhi
4   1931     Man of the Year        Pierre Laval
..   ...                 ...                 ...
86  2012  Person of the Year        Barack Obama
87  2013  Person of the Year        Pope Francis
88  2014  Person of the Year  The Ebola Fighters
89  2015  Person of the Year       Angela Merkel
90  2016  Person of the Year        Donald Trump

[91 rows x 3 columns]

Here is the version where we specify, by name, which columns to **remove**

In [17]:
time.drop(columns=['Year', 'Honor', 'Name'])

Country  Birth Year  Death Year  \
0   United States      1902.0      1974.0   
1   United States      1875.0      1940.0   
2   United States      1874.0      1962.0   
3           India      1869.0      1948.0   
4          France      1883.0      1945.0   
..            ...         ...         ...   
86  United States      1961.0         NaN   
87   Vatican City      1936.0         NaN   
88            NaN         NaN         NaN   
89        Germany      1954.0         NaN   
90  United States      1946.0         NaN   

                                                Title    Category  \
0                                   US Air Mail Pilot         NaN   
1                                 Founder of Chrysler   Economics   
2   Member of the German Reparations International...   Diplomacy   
3                                                 NaN  Revolution   
4                            Prime Minister of France    Politics   
..                                                ...         ...   
86                     President of the United States    Politics   
87                  Pope of the Roman Catholic Church    Religion   
88                                                NaN     Science   
89                              Chancellor of Germany    Politics   
90                     President of the United States    Politics   

                                              Context  Category Ordinal  \
0                     First Solo Transatlantic Flight                 0   
1                               Chrysler/Dodge Merger                 7   
2                                          Young Plan                10   
3                                          Salt March                 4   
4                                                 NaN                 3   
..                                                ...               ...   
86                              Presidential Election                 3   
87                                     Papal Conclave                 5   
88                                     Ebola Epidemic                12   
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...                 3   
90                              Presidential Election                 3   

    Lifespan  Birth Decade  Name Length Initials  
0       72.0        1900.0           17       CL  
1       65.0        1870.0           15       WC  
2       88.0        1870.0           13      ODY  
3       79.0        1860.0           14       MG  
4       62.0        1880.0           12       PL  
..       ...           ...          ...      ...  
86       NaN        1960.0           12       BO  
87       NaN        1930.0           12       PF  
88       NaN           NaN           18       TE  
89       NaN        1950.0           13       AM  
90       NaN        1940.0           12       DT  

[91 rows x 11 columns]

To apply the changes, assign the result back to the same variable `time`

In [18]:
time = time.drop(columns=['Year', 'Honor', 'Name'])
time

Country  Birth Year  Death Year  \
0   United States      1902.0      1974.0   
1   United States      1875.0      1940.0   
2   United States      1874.0      1962.0   
3           India      1869.0      1948.0   
4          France      1883.0      1945.0   
..            ...         ...         ...   
86  United States      1961.0         NaN   
87   Vatican City      1936.0         NaN   
88            NaN         NaN         NaN   
89        Germany      1954.0         NaN   
90  United States      1946.0         NaN   

                                                Title    Category  \
0                                   US Air Mail Pilot         NaN   
1                                 Founder of Chrysler   Economics   
2   Member of the German Reparations International...   Diplomacy   
3                                                 NaN  Revolution   
4                            Prime Minister of France    Politics   
..                                                ...         ...   
86                     President of the United States    Politics   
87                  Pope of the Roman Catholic Church    Religion   
88                                                NaN     Science   
89                              Chancellor of Germany    Politics   
90                     President of the United States    Politics   

                                              Context  Category Ordinal  \
0                     First Solo Transatlantic Flight                 0   
1                               Chrysler/Dodge Merger                 7   
2                                          Young Plan                10   
3                                          Salt March                 4   
4                                                 NaN                 3   
..                                                ...               ...   
86                              Presidential Election                 3   
87                                     Papal Conclave                 5   
88                                     Ebola Epidemic                12   
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...                 3   
90                              Presidential Election                 3   

    Lifespan  Birth Decade  Name Length Initials  
0       72.0        1900.0           17       CL  
1       65.0        1870.0           15       WC  
2       88.0        1870.0           13      ODY  
3       79.0        1860.0           14       MG  
4       62.0        1880.0           12       PL  
..       ...           ...          ...      ...  
86       NaN        1960.0           12       BO  
87       NaN        1930.0           12       PF  
88       NaN           NaN           18       TE  
89       NaN        1950.0           13       AM  
90       NaN        1940.0           12       DT  

[91 rows x 11 columns]

### Renaming Columns

You can do this using a `dictionary`

In [19]:
time = time.rename(columns={'Birth Year' : 'Born',
                        'Death Year' : 'Died'})
time

Country    Born    Died  \
0   United States  1902.0  1974.0   
1   United States  1875.0  1940.0   
2   United States  1874.0  1962.0   
3           India  1869.0  1948.0   
4          France  1883.0  1945.0   
..            ...     ...     ...   
86  United States  1961.0     NaN   
87   Vatican City  1936.0     NaN   
88            NaN     NaN     NaN   
89        Germany  1954.0     NaN   
90  United States  1946.0     NaN   

                                                Title    Category  \
0                                   US Air Mail Pilot         NaN   
1                                 Founder of Chrysler   Economics   
2   Member of the German Reparations International...   Diplomacy   
3                                                 NaN  Revolution   
4                            Prime Minister of France    Politics   
..                                                ...         ...   
86                     President of the United States    Politics   
87                  Pope of the Roman Catholic Church    Religion   
88                                                NaN     Science   
89                              Chancellor of Germany    Politics   
90                     President of the United States    Politics   

                                              Context  Category Ordinal  \
0                     First Solo Transatlantic Flight                 0   
1                               Chrysler/Dodge Merger                 7   
2                                          Young Plan                10   
3                                          Salt March                 4   
4                                                 NaN                 3   
..                                                ...               ...   
86                              Presidential Election                 3   
87                                     Papal Conclave                 5   
88                                     Ebola Epidemic                12   
89  Debt Crisis; Refugee Crisis; Paris Terrorist A...                 3   
90                              Presidential Election                 3   

    Lifespan  Birth Decade  Name Length Initials  
0       72.0        1900.0           17       CL  
1       65.0        1870.0           15       WC  
2       88.0        1870.0           13      ODY  
3       79.0        1860.0           14       MG  
4       62.0        1880.0           12       PL  
..       ...           ...          ...      ...  
86       NaN        1960.0           12       BO  
87       NaN        1930.0           12       PF  
88       NaN           NaN           18       TE  
89       NaN        1950.0           13       AM  
90       NaN        1940.0           12       DT  

[91 rows x 11 columns]

## Missing Data

### Fundamentals of missing data

In the previous example, you saw that some entries had `Nan` to fill in for **missing data**

Missing data is one of the most recurrent problems in data analysis

In fact, there are a lot of studies and methods on how to handle it!

We will work with a small dataset called [missing.csv](https://www.dropbox.com/s/5h4k4rszebd6p0n/missing.csv?raw=1) (it's the same as data.csv from last week, but now with missing values)

In [20]:
df = pd.read_csv('https://www.dropbox.com/s/5h4k4rszebd6p0n/missing.csv?raw=1')
df

Name   Age   Height
0       Nick  21.0     1.85
1      Chris  29.0  Unknown
2        Tim  28.0     1.75
3        Ron   NaN     1.81
4     Monica  35.0  Unknown
5  Cassandra  21.0     1.66

The first thing to notice is that the second column is being shown with decimals, even though they were integers in the original file

Moreover, the blank entry is displayed as `NaN`

`NaN` stands for *Not A Number* as is part of the floating point specification, typically used when a calculation has no valid numerical result

In fact, Python considers `NaN` as a floating point (or `float64`)

In [21]:
df

Name   Age   Height
0       Nick  21.0     1.85
1      Chris  29.0  Unknown
2        Tim  28.0     1.75
3        Ron   NaN     1.81
4     Monica  35.0  Unknown
5  Cassandra  21.0     1.66

In [22]:
type(df.at[3, 'Age'])

numpy.float64

As a result, the entire column has been turned into `float64`, instead of `int64`

 This is why it was being shown as decimals!

In [23]:
df['Age']

0    21.0
1    29.0
2    28.0
3     NaN
4    35.0
5    21.0
Name: Age, dtype: float64

### Incorrectly Imported Missing Data

The missing value in the `Age` column was generated because the cell was blank

By default, `Pandas` will treat certain cell values, such as a *blank*, `NULL`, `NaN`, and `n/a` as missing values

The `Height` column contains some entries which we recognise as missing data (i.e. `Unknown`) but that Pandas did not

In fact, the entire column has been imported as strings, not numbers!

In [24]:
df

Name   Age   Height
0       Nick  21.0     1.85
1      Chris  29.0  Unknown
2        Tim  28.0     1.75
3        Ron   NaN     1.81
4     Monica  35.0  Unknown
5  Cassandra  21.0     1.66

In [25]:
print(df.at[1, 'Height'])
print(type(df.at[1, 'Height']))

Unknown
<class 'str'>


This in turn made `Pandas` believe that the whole column has strings!

In [26]:
print(df.at[0, 'Height'])
print(type(df.at[0, 'Height']))

1.85
<class 'str'>


What happens if we multiply this column by `100`?

In [27]:
df['Height times 100'] = df['Height'] * 100
df

Name   Age   Height                                   Height times 100
0       Nick  21.0     1.85  1.851.851.851.851.851.851.851.851.851.851.851....
1      Chris  29.0  Unknown  UnknownUnknownUnknownUnknownUnknownUnknownUnkn...
2        Tim  28.0     1.75  1.751.751.751.751.751.751.751.751.751.751.751....
3        Ron   NaN     1.81  1.811.811.811.811.811.811.811.811.811.811.811....
4     Monica  35.0  Unknown  UnknownUnknownUnknownUnknownUnknownUnknownUnkn...
5  Cassandra  21.0     1.66  1.661.661.661.661.661.661.661.661.661.661.661....

We need to tell Pandas in advance to treat `Unknown` not as a string, but as a missing value

Lets re-import the data set passing a list of values which represent missing values as the `na_values` parameter

In [28]:
df = pd.read_csv('https://www.dropbox.com/s/5h4k4rszebd6p0n/missing.csv?raw=1', 
                 na_values=['Unknown'])
df

Name   Age  Height
0       Nick  21.0    1.85
1      Chris  29.0     NaN
2        Tim  28.0    1.75
3        Ron   NaN    1.81
4     Monica  35.0     NaN
5  Cassandra  21.0    1.66

Now the column is `float64`

In [29]:
df['Height']

0    1.85
1     NaN
2    1.75
3    1.81
4     NaN
5    1.66
Name: Height, dtype: float64

Multiplying `Height` now works as expected.

In [30]:
df['Height times 100'] = df['Height'] * 100
df

Name   Age  Height  Height times 100
0       Nick  21.0    1.85             185.0
1      Chris  29.0     NaN               NaN
2        Tim  28.0    1.75             175.0
3        Ron   NaN    1.81             181.0
4     Monica  35.0     NaN               NaN
5  Cassandra  21.0    1.66             166.0

In [31]:
# Another example
df['Test'] = df['Age'] + df['Height']
df

Name   Age  Height  Height times 100   Test
0       Nick  21.0    1.85             185.0  22.85
1      Chris  29.0     NaN               NaN    NaN
2        Tim  28.0    1.75             175.0  29.75
3        Ron   NaN    1.81             181.0    NaN
4     Monica  35.0     NaN               NaN    NaN
5  Cassandra  21.0    1.66             166.0  22.66

### Methods to address missing data

As mentioned before, there are different ways to deal with missing data

The method you should use depends on your analysis, the application domain and what you want to achieve

Here are some common methods:

    1. Leave the missing values as NaN
    2. Lookup the values from another data source
    3. Delete the rows / observations
    4. Compute reasonable guesses about the values

In **option 1**, it is often reasonable to just leave the data as missing data

For example, if we wanted to calculate average `Height`, it may not matter that we are missing `Age` for one row

As for **option 2**, if we have some other ways of going back and getting correct data, we could just update the values with `df.at[3, 'Age'] = ...`

Let's assume we are going for **option 3** and want to remove any rows where we do no have `Age`

If we know the row we want to delete (`3`), we can use the `drop` method, only this time we specify rows, not columns. As in:

    df.drop(3)

However, often there are many missing values, and we don't want to check on the index of each

A good way to remove rows with missing data is to specify which rows to **keep**

The `isnull` method will filter for rows which contains `NaN`.

In [32]:
df[df['Age'].isnull()]

Name  Age  Height  Height times 100  Test
3  Ron  NaN    1.81             181.0   NaN

Of course, this is the *opposite* of specifying which to keep

The **notnull** method will show us which rows do not contain `NaN` in the specified column.

In [33]:
df[df['Age'].notnull()]

Name   Age  Height  Height times 100   Test
0       Nick  21.0    1.85             185.0  22.85
1      Chris  29.0     NaN               NaN    NaN
2        Tim  28.0    1.75             175.0  29.75
4     Monica  35.0     NaN               NaN    NaN
5  Cassandra  21.0    1.66             166.0  22.66

Notice that the filter is only checking the column we specified `Age`, not the other column with `NaN` values

We can apply the filter by assigning the result back to `df`

In [34]:
df = df[df['Age'].notnull()]
df

Name   Age  Height  Height times 100   Test
0       Nick  21.0    1.85             185.0  22.85
1      Chris  29.0     NaN               NaN    NaN
2        Tim  28.0    1.75             175.0  29.75
4     Monica  35.0     NaN               NaN    NaN
5  Cassandra  21.0    1.66             166.0  22.66

Since the `Age` column no longer contains any non-integers, it is safe to convert the column to `int64` type

We can do this with the `astype` method

In [35]:
df = df.astype({'Age': 'int64'})
df

Name  Age  Height  Height times 100   Test
0       Nick   21    1.85             185.0  22.85
1      Chris   29     NaN               NaN    NaN
2        Tim   28    1.75             175.0  29.75
4     Monica   35     NaN               NaN    NaN
5  Cassandra   21    1.66             166.0  22.66

### Updating Indices

After these operations, the current data frame has the index 0, 1, 2, 4, 5

We could leave it like this, being careful to select the correct rows with `loc`. However, we can also reset the index with `reset_index`

Note that we need to specify `drop=True`, without this we will end up with an extra column with the old index

In [36]:
df = df.reset_index(drop=True)
df

Name  Age  Height  Height times 100   Test
0       Nick   21    1.85             185.0  22.85
1      Chris   29     NaN               NaN    NaN
2        Tim   28    1.75             175.0  29.75
3     Monica   35     NaN               NaN    NaN
4  Cassandra   21    1.66             166.0  22.66

Alternatively, we can set the index to a particular column if we don't want default indexing

In [37]:
df = df.set_index('Name')
df

Age  Height  Height times 100   Test
Name                                           
Nick        21    1.85             185.0  22.85
Chris       29     NaN               NaN    NaN
Tim         28    1.75             175.0  29.75
Monica      35     NaN               NaN    NaN
Cassandra   21    1.66             166.0  22.66

Say we want to go with **option 4** and replace missing `Height` with a computed value

In this example we simply calculate the mean height of other people in the data set (to the nearest cm)

**Note:** The `mean` method just ignores the missing data and gives us the average of the non-missing data, which is what we want

In [38]:
mean_height = df['Height'].mean()

mean_height

1.7533333333333332

In [39]:
mean_height = round(mean_height, 
                    2)
mean_height

1.75

Of course there are many more sophisticated ways to predict a replacement value!

For instance, could fit a regression with `Age`

Or if we had categories such as `Gender` we could group by categories and do predictions for each

In this example we will just use `1.75` for every missing value.

In [40]:
df = df.fillna({'Height' : mean_height})
df

Age  Height  Height times 100   Test
Name                                           
Nick        21    1.85             185.0  22.85
Chris       29    1.75               NaN    NaN
Tim         28    1.75             175.0  29.75
Monica      35    1.75               NaN    NaN
Cassandra   21    1.66             166.0  22.66

##  Joints and Aggregations

Let's create two small dataset to use as example

In [41]:
books = pd.DataFrame({'Author' : ['J. R. R. Tolkien',
                                  'George R. R. Martin',
                                  'J. K. Rowling', 
                                  'Suzanne Collins']},
                     index = ['The Lord of the Rings',
                              'Game of Thrones',
                              'Harry Potter',
                              'The Hunger Games'])
books

Author
The Lord of the Rings     J. R. R. Tolkien
Game of Thrones        George R. R. Martin
Harry Potter                 J. K. Rowling
The Hunger Games           Suzanne Collins

In [42]:
films = pd.DataFrame({'Year of First Film' : [1999, 2001, 2001, 2012],
                      'Number of Films' : [3, 2, 8, 4]},
                     index = ['The Matrix',
                              'The Lord of the Rings',
                              'Harry Potter',
                              'The Hunger Games'])


films

Year of First Film  Number of Films
The Matrix                           1999                3
The Lord of the Rings                2001                2
Harry Potter                         2001                8
The Hunger Games                     2012                4

The `.join` operation is used to join the columns of two different data sets based on matching index

In [43]:
books.join(films)

Author  Year of First Film  \
The Lord of the Rings     J. R. R. Tolkien              2001.0   
Game of Thrones        George R. R. Martin                 NaN   
Harry Potter                 J. K. Rowling              2001.0   
The Hunger Games           Suzanne Collins              2012.0   

                       Number of Films  
The Lord of the Rings              2.0  
Game of Thrones                    NaN  
Harry Potter                       8.0  
The Hunger Games                   4.0

### Types of Join

In the previous example, the `books` data set is the *'left'* dataset and `films` is the *'right'* dataset

A left join keeps all of the data from the *'left'* data set and adds in the applicable data from the *'right'* data set where the keys match up

If there is no film, the cells are populated with `NaN` (e.g. Game of Thrones)

The above operations is equivalent to

    books.join(films, how='left')

We can also do a right join. This keeps all of the films, adding the book data where applicable. If there is no book, the cells are populated with `NaN`.

In [44]:
books.join(films, how='right')

Author  Year of First Film  Number of Films
The Matrix                          NaN                1999                3
The Lord of the Rings  J. R. R. Tolkien                2001                2
Harry Potter              J. K. Rowling                2001                8
The Hunger Games        Suzanne Collins                2012                4

This is almost equivalent to `films.join(books)` (or `films.join(books, how='left')`), with the exception of the order in which the columns appear

If we want to keep all of the data (book **OR** film), we can use an outer join

In [45]:
films.join(books, how='outer')

Year of First Film  Number of Films  \
Game of Thrones                       NaN              NaN   
Harry Potter                       2001.0              8.0   
The Hunger Games                   2012.0              4.0   
The Lord of the Rings              2001.0              2.0   
The Matrix                         1999.0              3.0   

                                    Author  
Game of Thrones        George R. R. Martin  
Harry Potter                 J. K. Rowling  
The Hunger Games           Suzanne Collins  
The Lord of the Rings     J. R. R. Tolkien  
The Matrix                             NaN

And if we only want to keep the data (book **AND** film), we can use an inner join

In [46]:
films.join(books, how='inner')

Year of First Film  Number of Films            Author
The Lord of the Rings                2001                2  J. R. R. Tolkien
Harry Potter                         2001                8     J. K. Rowling
The Hunger Games                     2012                4   Suzanne Collins

You should choose the join type based on what your resulting data table is intended to describe

For instance, the inner join gave us a table of films based on books

Contrarily, the left join gave us a list of books with additional information on the film (if any)

Once again, remember that `.join` is not modifying the dataframe, so if you want to save the result, assign it to either the same or a different variable with an appropriate name

    films = films.join(books, how='left')

    books = films.join(books, how='right')
    
    films_based_on_books = films.join(books, how='inner')
    
    favourite_series = films.join(books, how='outer')

This table may be used as a reminder of the difference between the joins


| Type of Join   | Keeps Rows of Left Data | Keeps Rows of Right Data |
| :------------- | ----------------------: | -----------------------: |
| left (default) | yes                     | only if matching left    |
| right          | only if matching right  | yes                      |
| outer          | yes                     | yes                      |
| inner          | only if matching right  | only if matching left    |

### Joining Different Columns

`.join` joins by comparing indexes of each dataframe

Sometimes the key column(s) is not the index (particularly if you are using default indexing)

If you need to join based on columns other than the index, you should use `merge`

For example, it is possible that we would encounter data with default indexes as follows:

In [47]:
books = books.reset_index()
books = books.rename(columns={'index' : 'Book Series Title'})
books

Book Series Title               Author
0  The Lord of the Rings     J. R. R. Tolkien
1        Game of Thrones  George R. R. Martin
2           Harry Potter        J. K. Rowling
3       The Hunger Games      Suzanne Collins

In [48]:
films = films.reset_index()
films = films.rename(columns={'index' : 'Film Series Title'})
films

Film Series Title  Year of First Film  Number of Films
0             The Matrix                1999                3
1  The Lord of the Rings                2001                2
2           Harry Potter                2001                8
3       The Hunger Games                2012                4

If we join on the index, the result is nonsense!

In [49]:
books.join(films) # WRONG

Book Series Title               Author      Film Series Title  \
0  The Lord of the Rings     J. R. R. Tolkien             The Matrix   
1        Game of Thrones  George R. R. Martin  The Lord of the Rings   
2           Harry Potter        J. K. Rowling           Harry Potter   
3       The Hunger Games      Suzanne Collins       The Hunger Games   

   Year of First Film  Number of Films  
0                1999                3  
1                2001                2  
2                2001                8  
3                2012                4

We could change the `Book Series Title` and `Film Series Title` to indexes and join with `.join`

Or we can use `.merge`, in which left, right, outer, and inner joins work the same way

However, it is not the index we are comparing, it is the column specified with `left_on=` and `right_on=`

In [50]:
books.merge(films,
            how='inner',
            left_on='Book Series Title',
            right_on='Film Series Title')

Book Series Title            Author      Film Series Title  \
0  The Lord of the Rings  J. R. R. Tolkien  The Lord of the Rings   
1           Harry Potter     J. K. Rowling           Harry Potter   
2       The Hunger Games   Suzanne Collins       The Hunger Games   

   Year of First Film  Number of Films  
0                2001                2  
1                2001                8  
2                2012                4

### Data Aggregation

Before continuing, we will re set the index of the books

In [51]:
books = books.set_index('Book Series Title', drop=True)
books

Author
Book Series Title                         
The Lord of the Rings     J. R. R. Tolkien
Game of Thrones        George R. R. Martin
Harry Potter                 J. K. Rowling
The Hunger Games           Suzanne Collins

Let's import a dataset of a list of books

In [52]:
volumes = pd.read_csv('https://www.dropbox.com/s/9flqjjvetgbex97/volumes.csv?raw=1')
volumes

Series                                      Title  Rating  \
0            Harry Potter   Harry Potter and the Philosopher's Stone    4.47   
1            Harry Potter    Harry Potter and the Chamber of Secrets    4.42   
2            Harry Potter   Harry Potter and the Prisoner of Azkaban    4.56   
3            Harry Potter        Harry Potter and the Goblet of Fire    4.55   
4            Harry Potter  Harry Potter and the Order of the Phoenix    4.49   
5            Harry Potter     Harry Potter and the Half-Blood Prince    4.57   
6            Harry Potter       Harry Potter and the Deathly Hallows    4.61   
7   The Lord of the Rings                 The Fellowship of the Ring    4.36   
8   The Lord of the Rings                             The Two Towers    4.44   
9   The Lord of the Rings                     The Return of the King    4.53   
10        Game of Thrones                          A Game of Thrones    4.45   
11        Game of Thrones                           A Clash of Kings    4.41   
12        Game of Thrones                          A Storm of Swords    4.54   
13        Game of Thrones                          A Feast for Crows    4.13   
14        Game of Thrones                       A Dance with Dragons    4.33   
15       The Hunger Games                           The Hunger Games    4.33   
16       The Hunger Games                              Catching Fire    4.29   
17       The Hunger Games                                 Mockingjay    4.03   

    Year  
0   1997  
1   1998  
2   1999  
3   2000  
4   2003  
5   2005  
6   2007  
7   1954  
8   1954  
9   1955  
10  1996  
11  1998  
12  2000  
13  2005  
14  2011  
15  2008  
16  2009  
17  2010

We want to summarise by series, the `.groupby` method gives you the name of the column which has the groups

The result is a Python object which we will use for the next step

In [53]:
groups = volumes.groupby('Series')
groups

We will use `count` and `mean` to work out the number of books and the average rating, and use `min` to work out the first publication year.

Other operations available include `sum` and `max` (you can check others [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html))

Now we have our data summary by groups

In [54]:
summary = groups.agg({'count', 'mean', 'min'})
summary

Rating                     Year            
                           mean   min count         mean   min count
Series                                                              
Game of Thrones        4.372000  4.13     5  2002.000000  1996     5
Harry Potter           4.524286  4.42     7  2001.285714  1997     7
The Hunger Games       4.216667  4.03     3  2009.000000  2008     3
The Lord of the Rings  4.443333  4.36     3  1954.333333  1954     3

### Joining Aggregation Data

We now have a dataframe with two sub-frames (one for `Rating` and one for `Year`) we can easily separate them

In [55]:
rating_summary = summary['Rating']
rating_summary

mean   min  count
Series                                      
Game of Thrones        4.372000  4.13      5
Harry Potter           4.524286  4.42      7
The Hunger Games       4.216667  4.03      3
The Lord of the Rings  4.443333  4.36      3

In [56]:
year_summary = summary['Year']
year_summary

mean   min  count
Series                                         
Game of Thrones        2002.000000  1996      5
Harry Potter           2001.285714  1997      7
The Hunger Games       2009.000000  2008      3
The Lord of the Rings  1954.333333  1954      3

Let's rename the columns for the `Ratings` aggregation, and remove the unneeded `min` column

In [57]:
rating_summary = summary['Rating'].rename(
    columns={'mean' : 'Average Rating',
             'count' : 'Number of Books'})
rating_summary = rating_summary.drop(columns={'min'})
rating_summary

Average Rating  Number of Books
Series                                                
Game of Thrones              4.372000                5
Harry Potter                 4.524286                7
The Hunger Games             4.216667                3
The Lord of the Rings        4.443333                3

We may also want to round off the average ratings

In [58]:
rating_summary['Average Rating'] = rating_summary['Average Rating'].round(2)

rating_summary

Average Rating  Number of Books
Series                                                
Game of Thrones                  4.37                5
Harry Potter                     4.52                7
The Hunger Games                 4.22                3
The Lord of the Rings            4.44                3

Let's also rename the column from the `Year` aggregation and drop the other columns

In [59]:
year_summary = year_summary.rename(columns={'min' : 'First Published'})
year_summary = year_summary.drop(columns={'mean', 'count'})
year_summary

First Published
Series                                
Game of Thrones                   1996
Harry Potter                      1997
The Hunger Games                  2008
The Lord of the Rings             1954

Now we can join the `books`, `rating_summary` and `year_summary` dataframes

Since all have the same keys we don't need to worry about join type, but this is a left join so will keep everything in the `books` data frame if it didn't match

In [60]:
books.join(rating_summary).join(year_summary)

Author  Average Rating  Number of Books  \
Book Series Title                                                             
The Lord of the Rings     J. R. R. Tolkien            4.44                3   
Game of Thrones        George R. R. Martin            4.37                5   
Harry Potter                 J. K. Rowling            4.52                7   
The Hunger Games           Suzanne Collins            4.22                3   

                       First Published  
Book Series Title                       
The Lord of the Rings             1954  
Game of Thrones                   1996  
Harry Potter                      1997  
The Hunger Games                  2008

## Melt and Pivoting

### Wide vs Long Data (some real-life examples)

Wide data contains a column for each variable, and a row for each entity

The "entity" ID (in this case `Name`, but it could be an ID number etc.) is in the first column, or could be the index

| Name    | Age  | Height | Hair Colour |
| ------: | ---: | -----: | :---------- |
| Alice   |   36 |  1.68  | Blonde      |
| Bob     |   28 |  1.73  | Red         |
| Charlie |   29 |  1.60  | -           |

Long data contains a row for each observation of a variable

his is also called entity-attribute-value data

Note that rows can be omitted if there is missing data

| Entity ID | Attribute / Variable   |   Value |
| --------: | :--------------------- | ------: |
| Alice     | Age                    |      36 |
| Bob       | Age                    |      28 |
| Charlie   | Age                    |      29 |
| Alice     | Height                 |    1.68 |
| Bob       | Height                 |    1.73 |
| Charlie   | Height                 |    1.60 |
| Alice     | Hair Colour            |  Blonde |
| Bob       | Hair Colour            |     Red |

### Tidy Data

Defined by Hadley Wickham in [this article](https://vita.had.co.nz/papers/tidy-data.pdf), it describes long and wide data, and gives more examples on how to better work with both

It is written for R users!

### Melting Wide to Long Data
![Fig. 1](https://www.dropbox.com/s/giskupj9ibff4bd/fig1.jpg?raw=1)

Once again, let's import a new dataset

In [61]:
stock = pd.read_csv('https://www.dropbox.com/s/dl0bz061v5biv4k/stock.csv?raw=1')
stock

Company Symbol  1980  1990   2000    2010     2020
0      Apple   AAPL  0.51  1.22   3.88   37.53   318.73
1     Google  GOOGL   NaN   NaN    NaN  312.54  1518.73
2  Microsoft   MSFT   NaN  1.03  53.31   28.21   185.38

**What kind of data is this?**

**What are the entities, and which columns are ID columns?**

**What takes the place of attributes (variables) in this case?**

Let's melt this data using the `.melt` method

In [62]:
stock.melt(id_vars = ['Company', 'Symbol'])

Company Symbol variable    value
0       Apple   AAPL     1980     0.51
1      Google  GOOGL     1980      NaN
2   Microsoft   MSFT     1980      NaN
3       Apple   AAPL     1990     1.22
4      Google  GOOGL     1990      NaN
5   Microsoft   MSFT     1990     1.03
6       Apple   AAPL     2000     3.88
7      Google  GOOGL     2000      NaN
8   Microsoft   MSFT     2000    53.31
9       Apple   AAPL     2010    37.53
10     Google  GOOGL     2010   312.54
11  Microsoft   MSFT     2010    28.21
12      Apple   AAPL     2020   318.73
13     Google  GOOGL     2020  1518.73
14  Microsoft   MSFT     2020   185.38

We can ensure that the new columns get named correctly by using the `var_name` and `value_name` options

In [63]:
stock = stock.melt(id_vars = ['Company', 'Symbol'], var_name='Year', value_name='Price (USD)')
stock

Company Symbol  Year  Price (USD)
0       Apple   AAPL  1980         0.51
1      Google  GOOGL  1980          NaN
2   Microsoft   MSFT  1980          NaN
3       Apple   AAPL  1990         1.22
4      Google  GOOGL  1990          NaN
5   Microsoft   MSFT  1990         1.03
6       Apple   AAPL  2000         3.88
7      Google  GOOGL  2000          NaN
8   Microsoft   MSFT  2000        53.31
9       Apple   AAPL  2010        37.53
10     Google  GOOGL  2010       312.54
11  Microsoft   MSFT  2010        28.21
12      Apple   AAPL  2020       318.73
13     Google  GOOGL  2020      1518.73
14  Microsoft   MSFT  2020       185.38

The `NaN` values aren't really contributing anything and are just there because they were in the wide data set, so let's remove them!

In [64]:
stock = stock[stock['Price (USD)'].notnull()]
stock

Company Symbol  Year  Price (USD)
0       Apple   AAPL  1980         0.51
3       Apple   AAPL  1990         1.22
5   Microsoft   MSFT  1990         1.03
6       Apple   AAPL  2000         3.88
8   Microsoft   MSFT  2000        53.31
9       Apple   AAPL  2010        37.53
10     Google  GOOGL  2010       312.54
11  Microsoft   MSFT  2010        28.21
12      Apple   AAPL  2020       318.73
13     Google  GOOGL  2020      1518.73
14  Microsoft   MSFT  2020       185.38

If you prefer the data sorted by entity, then variable, you can re-sort

In [65]:
stock.sort_values(['Symbol', 'Year'])

Company Symbol  Year  Price (USD)
0       Apple   AAPL  1980         0.51
3       Apple   AAPL  1990         1.22
6       Apple   AAPL  2000         3.88
9       Apple   AAPL  2010        37.53
12      Apple   AAPL  2020       318.73
10     Google  GOOGL  2010       312.54
13     Google  GOOGL  2020      1518.73
5   Microsoft   MSFT  1990         1.03
8   Microsoft   MSFT  2000        53.31
11  Microsoft   MSFT  2010        28.21
14  Microsoft   MSFT  2020       185.38

Once again, Python has not detected that the columns were integer types

In [66]:
stock[stock['Year'] >= 2000]  # Error!

TypeError: '>=' not supported between instances of 'str' and 'int'

As you can see, the data type is `object`, meaning these numbers are stored as `str`

In [67]:
stock['Year']

0     1980
3     1990
5     1990
6     2000
8     2000
9     2010
10    2010
11    2010
12    2020
13    2020
14    2020
Name: Year, dtype: object

The columns which was originally headings has become `object` (`str`) types

We can change the data type of the `Year` column using `astype`

In [68]:
stock = stock.astype({'Year' : 'int64'})
stock

Company Symbol  Year  Price (USD)
0       Apple   AAPL  1980         0.51
3       Apple   AAPL  1990         1.22
5   Microsoft   MSFT  1990         1.03
6       Apple   AAPL  2000         3.88
8   Microsoft   MSFT  2000        53.31
9       Apple   AAPL  2010        37.53
10     Google  GOOGL  2010       312.54
11  Microsoft   MSFT  2010        28.21
12      Apple   AAPL  2020       318.73
13     Google  GOOGL  2020      1518.73
14  Microsoft   MSFT  2020       185.38

Now we are able to do numerical comparisons on the `Year` column

In [69]:
stock[stock['Year'] >= 2000]

Company Symbol  Year  Price (USD)
6       Apple   AAPL  2000         3.88
8   Microsoft   MSFT  2000        53.31
9       Apple   AAPL  2010        37.53
10     Google  GOOGL  2010       312.54
11  Microsoft   MSFT  2010        28.21
12      Apple   AAPL  2020       318.73
13     Google  GOOGL  2020      1518.73
14  Microsoft   MSFT  2020       185.38

### Pivoting Long to Wide Data
![Fig. 2](https://www.dropbox.com/s/x2i8xhzt0yvfip5/fig2.gif?raw=1)

Another new dataset...

In [70]:
weather = pd.read_csv('https://www.dropbox.com/s/1pyru339tll1njf/weather-canada.csv?raw=1')
weather

Station Name Province  Year  Mean Temperature (C)  \
0                     BEAR CREEK       BC  1971                  15.4   
1      COWICHAN BAY CHERRY POINT       BC  1971                  17.4   
2         COWICHAN LAKE FORESTRY       BC  1971                  18.8   
3          COWICHAN LAKE VILLAGE       BC  1971                  17.7   
4                DUNCAN FORESTRY       BC  1971                  17.7   
...                          ...      ...   ...                   ...   
81757                    GOOSE A       NL  2017                  15.8   
81758             HOPEDALE (AUT)       NL  2017                  11.6   
81759           MARY'S HARBOUR A       NL  2017                  14.5   
81760                       NAIN       NL  2017                  10.6   
81761                   WABUSH A       NL  2017                  12.9   

       Total Precipitation (mm)  
0                          20.9  
1                          12.8  
2                          21.3  
3                          36.4  
4                          18.1  
...                         ...  
81757                     109.0  
81758                      83.2  
81759                      56.9  
81760                      38.3  
81761                     129.0  

[81762 rows x 5 columns]

This was long data, however we have multiple variables per observation

We can pivot the table to see each station with the entry corresponding to different years

In [71]:
weather_p = weather.pivot(index='Station Name',
                          columns='Year',
                          values=['Mean Temperature (C)', 
                                  'Total Precipitation (mm)'])
weather_p

Mean Temperature (C)                                \
Year                                     1971  1972  1973  1974  1975  1976   
Station Name                                                                  
(AE) BOW SUMMIT                           NaN   NaN   NaN   NaN   NaN   NaN   
100 MILE HOUSE                           16.0  15.1  15.1  13.7  17.2  14.5   
100 MILE HOUSE 6NE                        NaN   NaN   NaN   NaN   NaN   NaN   
108 MILE HOUSE                            NaN   NaN  15.9   NaN   NaN   NaN   
108 MILE HOUSE ABEL LAKE                  NaN   NaN   NaN   NaN   NaN   NaN   
...                                       ...   ...   ...   ...   ...   ...   
YOYO                                      NaN   NaN   NaN   NaN   NaN   NaN   
ZAMA LO                                  15.7  14.3  15.3  13.5  17.5  14.3   
ZEBALLOS MURAUDE CREEK                    NaN   NaN   NaN   NaN   NaN   NaN   
ZEHNER                                    NaN   NaN   NaN   NaN   NaN   NaN   
ZHODA                                     NaN   NaN   NaN   NaN   NaN   NaN   

                                                  ...  \
Year                      1977  1978  1979  1980  ...   
Station Name                                      ...   
(AE) BOW SUMMIT            NaN   NaN   NaN   NaN  ...   
100 MILE HOUSE            13.5  15.5  16.0  14.1  ...   
100 MILE HOUSE 6NE         NaN   NaN   NaN   NaN  ...   
108 MILE HOUSE             NaN   NaN   NaN   NaN  ...   
108 MILE HOUSE ABEL LAKE   NaN   NaN   NaN   NaN  ...   
...                        ...   ...   ...   ...  ...   
YOYO                       NaN   NaN   NaN   NaN  ...   
ZAMA LO                   13.6  14.6  18.1  16.0  ...   
ZEBALLOS MURAUDE CREEK     NaN   NaN   NaN   NaN  ...   
ZEHNER                     NaN   NaN   NaN   NaN  ...   
ZHODA                      NaN   NaN   NaN   NaN  ...   

                         Total Precipitation (mm)                           \
Year                                         2008  2009  2010   2011  2012   
Station Name                                                                 
(AE) BOW SUMMIT                               NaN   NaN   NaN    NaN   NaN   
100 MILE HOUSE                                NaN   NaN   NaN    NaN   NaN   
100 MILE HOUSE 6NE                           48.4  47.0  21.4   75.8  68.4   
108 MILE HOUSE                                NaN   NaN   NaN    NaN   NaN   
108 MILE HOUSE ABEL LAKE                     37.6  42.8  12.0   55.4  57.2   
...                                           ...   ...   ...    ...   ...   
YOYO                                          NaN   NaN   NaN    NaN   NaN   
ZAMA LO                                      81.2  42.4  62.8  175.0   NaN   
ZEBALLOS MURAUDE CREEK                        NaN   NaN   NaN  123.2  56.4   
ZEHNER                                        NaN   NaN   NaN    NaN   NaN   
ZHODA                                         NaN   NaN   NaN    NaN   NaN   

                                                          
Year                      2013   2014  2015   2016  2017  
Station Name                                              
(AE) BOW SUMMIT            NaN    NaN   NaN    NaN   NaN  
100 MILE HOUSE             NaN    NaN   NaN    NaN   NaN  
100 MILE HOUSE 6NE        28.8  112.4  93.2  106.4   4.8  
108 MILE HOUSE             NaN    NaN   NaN    NaN   NaN  
108 MILE HOUSE ABEL LAKE  19.9   30.4  51.8   37.6   3.2  
...                        ...    ...   ...    ...   ...  
YOYO                       NaN    NaN   NaN    NaN   NaN  
ZAMA LO                    NaN    NaN   NaN    NaN   NaN  
ZEBALLOS MURAUDE CREEK     2.4   75.9  62.8   89.4  47.9  
ZEHNER                     NaN    NaN   NaN    NaN   NaN  
ZHODA                      NaN    NaN   NaN    NaN   NaN  

[4808 rows x 94 columns]

You can see that the method was capable of "grouping" data, either for mean temp or precipitations!

This allows us to create new dataframes based only on the required info

In [72]:
temperature = weather_p['Mean Temperature (C)']
temperature

Year                      1971  1972  1973  1974  1975  1976  1977  1978  \
Station Name                                                               
(AE) BOW SUMMIT            NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
100 MILE HOUSE            16.0  15.1  15.1  13.7  17.2  14.5  13.5  15.5   
100 MILE HOUSE 6NE         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
108 MILE HOUSE             NaN   NaN  15.9   NaN   NaN   NaN   NaN   NaN   
108 MILE HOUSE ABEL LAKE   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
...                        ...   ...   ...   ...   ...   ...   ...   ...   
YOYO                       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
ZAMA LO                   15.7  14.3  15.3  13.5  17.5  14.3  13.6  14.6   
ZEBALLOS MURAUDE CREEK     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
ZEHNER                     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
ZHODA                      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

Year                      1979  1980  ...  2008  2009  2010  2011  2012  2013  \
Station Name                          ...                                       
(AE) BOW SUMMIT            NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
100 MILE HOUSE            16.0  14.1  ...   NaN   NaN   NaN   NaN   NaN   NaN   
100 MILE HOUSE 6NE         NaN   NaN  ...  14.8  16.5  14.6  11.2  15.1  14.5   
108 MILE HOUSE             NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
108 MILE HOUSE ABEL LAKE   NaN   NaN  ...  16.1  18.7  16.8  14.4  18.0  17.5   
...                        ...   ...  ...   ...   ...   ...   ...   ...   ...   
YOYO                       NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
ZAMA LO                   18.1  16.0  ...  14.9  11.1   9.5  13.5   NaN   NaN   
ZEBALLOS MURAUDE CREEK     NaN   NaN  ...   NaN   NaN   NaN  13.9  16.4  18.4   
ZEHNER                     NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
ZHODA                      NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   

Year                      2014  2015  2016  2017  
Station Name                                      
(AE) BOW SUMMIT            NaN   NaN   NaN   NaN  
100 MILE HOUSE             NaN   NaN   NaN   NaN  
100 MILE HOUSE 6NE        16.6  16.5  14.9  17.3  
108 MILE HOUSE             NaN   NaN   NaN   NaN  
108 MILE HOUSE ABEL LAKE  18.5  18.0  16.7  18.0  
...                        ...   ...   ...   ...  
YOYO                       NaN   NaN   NaN   NaN  
ZAMA LO                    NaN   NaN   NaN   NaN  
ZEBALLOS MURAUDE CREEK    17.6  19.3  17.3  16.3  
ZEHNER                     NaN   NaN   NaN   NaN  
ZHODA                      NaN   NaN   NaN   NaN  

[4808 rows x 47 columns]

In [73]:
temp2010 = temperature[[2010]]
temp2010 = temp2010[temp2010[2010].notnull()]
temp2010

Year                      2010
Station Name                  
100 MILE HOUSE 6NE        14.6
108 MILE HOUSE ABEL LAKE  16.8
ABBOTSFORD A              18.3
ABEE AGDM                 15.3
ACADIA VALLEY             17.8
...                        ...
YOHIN                     17.7
YOHO NP OHARA LAKE        10.0
YOHO PARK                 12.0
YORKTON                   18.5
ZAMA LO                    9.5

[1323 rows x 1 columns]

In [74]:
temp2010.mean()

Year
2010    17.429932
dtype: float64

Let's pivot again, this time with the year as index

In [76]:
weather_p2 = weather.pivot(index='Year',
                           columns='Station Name',
                           values=['Mean Temperature (C)', 
                                   'Total Precipitation (mm)'])

weather_p2

Mean Temperature (C)                                    \
Station Name      (AE) BOW SUMMIT 100 MILE HOUSE 100 MILE HOUSE 6NE   
Year                                                                  
1971                          NaN           16.0                NaN   
1972                          NaN           15.1                NaN   
1973                          NaN           15.1                NaN   
1974                          NaN           13.7                NaN   
1975                          NaN           17.2                NaN   
1976                          NaN           14.5                NaN   
1977                          NaN           13.5                NaN   
1978                          NaN           15.5                NaN   
1979                          NaN           16.0                NaN   
1980                          NaN           14.1                NaN   
1981                          NaN           15.1                NaN   
1982                          NaN           14.8                NaN   
1983                          NaN           14.4                NaN   
1984                          NaN           14.7                NaN   
1985                          NaN           16.6                NaN   
1986                          NaN           13.5                NaN   
1987                          NaN           15.4               15.1   
1988                          NaN           14.9               14.7   
1989                          NaN           15.0               14.8   
1990                          NaN           16.3               16.0   
1991                          NaN           14.9               14.5   
1992                          NaN           15.7               15.3   
1993                          NaN           13.5               13.3   
1994                          NaN           16.7               16.6   
1995                          NaN           15.9               15.5   
1996                          NaN           15.3               15.4   
1997                          NaN           14.9               14.5   
1998                         12.1           18.1               17.4   
1999                          NaN           14.9               14.1   
2000                          9.1            NaN               14.6   
2001                          9.6            NaN               14.3   
2002                         10.0            NaN               15.2   
2003                         11.1            NaN               16.2   
2004                          NaN            NaN                NaN   
2005                          NaN            NaN               14.4   
2006                          NaN            NaN               16.8   
2007                          NaN            NaN               16.6   
2008                          NaN            NaN               14.8   
2009                          NaN            NaN               16.5   
2010                          NaN            NaN               14.6   
2011                          NaN            NaN               11.2   
2012                          NaN            NaN               15.1   
2013                          NaN            NaN               14.5   
2014                          NaN            NaN               16.6   
2015                          NaN            NaN               16.5   
2016                          NaN            NaN               14.9   
2017                          NaN            NaN               17.3   

                                                                        \
Station Name 108 MILE HOUSE 108 MILE HOUSE ABEL LAKE 150 MILE HOUSE 7N   
Year                                                                     
1971                    NaN                      NaN              15.5   
1972                    NaN                      NaN              14.1   
1973                   15.9                      NaN              14.1   
1974              

This allows us to get the mean of **all** stations!

In [77]:
weather_p2['Mean Temperature (C)'].mean()

Station Name
(AE) BOW SUMMIT             10.380000
100 MILE HOUSE              15.217241
100 MILE HOUSE 6NE          15.243333
108 MILE HOUSE              15.900000
108 MILE HOUSE ABEL LAKE    15.883333
                              ...    
YOYO                        13.350000
ZAMA LO                     15.392500
ZEBALLOS MURAUDE CREEK      17.028571
ZEHNER                      18.083333
ZHODA                       18.825000
Length: 4808, dtype: float64

### Pivoting with Aggregation

This next cell will give an error! **WHY**

In [78]:
weather.pivot(index='Province',
              columns='Year',
              values=['Mean Temperature (C)', 
                    'Total Precipitation (mm)'])  # Error!

ValueError: Index contains duplicate entries, cannot reshape

If we use `pivot_table` with `aggfunc`, we can tell Pandas what to do with these values for instance we may want the `mean`

In [79]:
weather.pivot_table(index='Province',
                    columns='Year',
                    values=['Mean Temperature (C)', 'Total Precipitation (mm)'],
                    aggfunc='mean')

Mean Temperature (C)                                              \
Year                     1971       1972       1973       1974       1975   
Province                                                                    
AB                  15.057655  13.122741  15.160550  14.393491  17.158160   
BC                  17.103300  15.994702  15.956347  14.695879  17.289267   
MB                  16.250000  16.147368  17.946281  20.249194  20.560504   
NB                  18.055172  17.520968  19.972581  17.357143  19.917742   
NL                  14.676596  14.482456  16.917241  13.165000  17.165517   
NS                  18.052174  17.608451  19.455844  16.082895  19.432877   
NT                  13.304545  11.672727  14.507143  13.164286  14.565789   
NU                   6.747222   4.506452   6.985294   7.532432   6.761765   
ON                  18.109699  18.820209  19.890492  19.444156  20.575974   
PE                  18.700000  18.006667  20.486667  17.133333  20.940000   
QC                  17.190986  17.606268  19.279911  17.842729  19.930769   
SK                  16.155233  15.250568  17.767895  19.021081  20.272826   
YT                  13.860000  12.914286  12.732258  11.845714  13.639474   

                                                                 ...  \
Year           1976       1977       1978       1979       1980  ...   
Province                                                         ...   
AB        15.045833  13.898171  15.229518  16.707165  15.106562  ...   
BC        14.965147  15.166482  17.168539  17.047714  15.647929  ...   
MB        18.756303  18.335897  17.756364  19.988288  18.857273  ...   
NB        17.650820  18.026230  18.147368  19.105357  17.536842  ...   
NL        14.661818  14.688235  15.235088  15.448276  13.847692  ...   
NS        17.419737  17.687500  17.256164  18.333803  17.115714  ...   
NT        13.507692  12.472000  11.266667  15.052000  12.096000  ...   
NU         6.460000   6.934375   4.663636   6.081818   6.170000  ...   
ON        18.717377  19.830201  19.069180  19.676431  19.504667  ...   
PE        18.200000  18.257143  18.371429  19.078571  17.371429  ...   
QC        17.253037  17.449657  17.867352  18.817401  17.412963  ...   
SK        18.302260  17.582065  17.468306  19.556425  18.334637  ...   
YT        12.808333  12.702941  13.840000  13.700000  13.000000  ...   

         Total Precipitation (mm)                                      \
Year                         2008        2009        2010        2011   
Province                                                                
AB                      57.499465   71.784946   63.541739   90.536782   
BC                      45.344048   33.367480   16.013248   70.216800   
MB                      91.120000   79.639394   76.880952   57.208197   
NB                      90.202941  150.363636  100.396552  126.335714   
NL                      66.025397  104.258333  122.348000  138.456522   
NS                      83.421429   93.652500   96.146875  103.189189   
NT                      29.204878   43.634483   44.984211   43.134483   
NU                      33.063889   29.728571   33.527027   24.541379   
ON                     103.389080   96.332143   88.679268   63.204516   
PE                      54.588889  139.633333  123.988889  121.822222   
QC                     129.135885  138.533659   87.888780   93.660101   
SK                      76.864211   73.309574   71.457143   74.325843   
YT                      62.766667   24.854167   48.369231   64.056000   

                                                                               
Year           2012        2013        2014        2015       2016       2017  
Province                                                                       
AB        72.519481   54.610132   46.419178   55.166038  77.486829  45.088500  
BC        46.386822   19.295000   43.874000   38.252863  50.727064  24.443781  
MB        61.190625   77.395238   43.877778   98.64310

## "Homework"

Look for data visualisations out there
    
    Plots from data/knowledge is beautiful
    Infographics
    Dashboards

Play/watch people play Among Us ([PC](https://store.steampowered.com/app/945360/Among_Us/) or [Mobile](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjUtar0xJ_sAhVhyoUKHZNlCToQFjACegQIAhAC&url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.innersloth.spacemafia%26hl%3Den_GB&usg=AOvVaw35aRiUcD6FcD9l3sgH3Ofy))

## An alternative way to run the dashboard for the Coursework Part 1
Thanks Pam!

## Lab (at last!)

I have created one notebook per topic

    Manipulating Columns
    Missing Values
    Joints & Aggregations
    Melt & Pivoting

Moreover, I have provided the guided versions (in .html)

Next week, I will release the Python solutions (.ipynb & html)

**NOTE**: All datasets can be accessed from the dropbox links, but can be also downloaded from Moodle if you prefer (including the ones used in the lecture)